### PR TITLE
New cmder.exe args and shared install capability

### PR DESCRIPTION
**Pull request by daxgames**
## [1.3.6-pre1](https://github.com/cmderdev/cmder) (2018-03-01)

The is a redo of #1665 - See the conversation there if you have questions.

**Fixed bugs:**

* Fixed Git version check recently added to master.

**Updates:**

* Modified Cmder tasks in default Conemu.xml to allow easily adding command line args for init.bat by adding some quotes.  This resulted in a ton of misc changes to this file.  See Adds below.
* Reworked `cmder.exe` command line argument handling to make it more flexible and easily added to.
* Reworked README.md tables to make them more readable in editors

**Implemented enhancements:**

* Append `%GIT_INSTALL_ROOT%\mingw32|64` folder to the path if it exists. Thanks: @gucong3000 
* Added `cmder.exe` command line args documenttion to `README.md`
* Added `:enhance_path` method to vendor\init.bat that modifies the path only if required.
   * To prepend: `call :enhance_path "%cmder_root%"`
   * to append: `call :enhance_path "%cmder_root%" append`
* Added `:enhance_path_recursive` method to vendor\init.bat that adds a path and all its sub directories to the path if required.  
   * Max recurse depth default is '1' configurable using `init.bat /max_depth [1-5]`. 6+ results in error. 
   * To prepend and go 3 levels deep: `call :enhance_path "%cmder_root%" 3`
   * To append and go 2 levels deep: `call :enhance_path "%cmder_root%" 2 append`
* Added ability to init.bat to accept command line args and documented them in README.md.  Allows users to change the behaviour of init.bat without editing the file.

| Argument                      | Description                                                                                      | Default                               |
| ----------------------------- | ----------------------------------------------------------------------------------------------   | ------------------------------------- |
| /c [user cmder root]          | Enables user bin and config folders for 'Cmder as admin' sessions due to non-shared environment. | not set                               |
| /d                            | Enables debug output.                                                                            | not set                               |
| /git_install_root [file path] | User specified Git installation root path.                                                       | '%CMDER_ROOT%\vendor\Git-for-Windows' |
| /home [home folder]           | User specified folder path to set `%HOME%` environment variable.                                 | '%userprofile%'                       |
| /max_depth [1-5]              | Define max recurse depth when adding to the path for `%cmder_root%\bin` and `%cmder_user_bin%`   | 1                                     |
| /svn_ssh [path to ssh.exe]    | Define %SVN_SSH% so we can use git svn with ssh svn repositories.                                | '%GIT_INSTALL_ROOT%\bin\ssh.exe'      |
| /user_aliases [file path]     | File path pointing to user aliases.                                                              | '%CMDER_ROOT%\config\user-liases.cmd' |
| /v                            | Enables verbose output.                                                                          | not set                               |

* Added new `cmder.exe /C \<path\>` argument
   * To use run Cmder.exe with "/C" command line argument. Example: `cmder.exe /C %userprofile%\cmder_config`
   * To use run with `Cmder as Admin` sessions you must specify "/c" command line argument to `init.bat` in tasks. See [README.md](./Readme.md) for details.
   * Enables shared Cmder install with Non-Portable Individual User Config
   * Supported by all supported shells (cmder, powershell, git bash, and external bash)
   * This will create the following directory structure if it is missing.

      ```
      c:\users\[username]\cmder_config
      ├───bin
      └───config
          └───profile.d
      ```

   * Shell init scripts run in the following order
      1. %cmder_root%\config\profile.d\*.[cmd|ps1|sh]
      1. %cmder_root%\config\user-profile.[cmd|ps1|sh]
      1. %userprofile%\cmder_config\config\profile.d\*.[cmd|ps1|sh]
      1. %userprofile%\cmder_config\config\user-profile.[cmd|ps1|sh]

### CODE REVIEW

**Comment by daxgames** [comment]
@cmderdev/trusted-contributors @DRSDavidSoft - This is a new PR based on #1665 because screwed that branch up somehow trying to rebase for @Stanzilla.  Please test the Nightly build of this commit.

File: launcher/src/CmderLauncher.cpp [hunk 1]
```diff
@@ -299,6 +316,116 @@ void UnregisterShellMenu(std::wstring opt, wchar_t* keyBaseName)
...
+};
+
+cmderOptions GetOption()
+{
+	cmderOptions cmderOptions;
+	LPWSTR *szArgList;
+	int argCount;
+
+	szArgList = CommandLineToArgvW(GetCommandLine(), &argCount);
+	/*
```
Comment (MartiUK) [inline launcher/src/CmderLauncher.cpp]: Any reason why these lines are left in commented out?

File: launcher/src/CmderLauncher.cpp [hunk 2]
```diff
@@ -299,6 +316,116 @@ void UnregisterShellMenu(std::wstring opt, wchar_t* keyBaseName)
...
+			}
+			else if (cmderOptions.cmderStart == L"" && PathFileExists(szArgList[i]))
+			{
+				cmderOptions.cmderStart = szArgList[i];
+			}
+			else {
+				MessageBox(NULL, L"Unrecognized parameter.\n\nValid options:\n\n    /c [CMDER User Root Path]\n\n    /task [ConEmu Task Name]\n\n    [/start [Start in Path] | [Start in Path]]\n\n    /single\n\nor\n\n    /register [USER | ALL]\n\nor\n\n    /unregister [USER | ALL]\n", MB_TITLE, MB_OK);
+				cmderOptions.error = true;
+			}
+		//}
```
Comment (MartiUK) [inline launcher/src/CmderLauncher.cpp]: Commented out brace?

File: launcher/src/CmderLauncher.cpp [hunk 3]
```diff
@@ -308,35 +435,25 @@ int APIENTRY _tWinMain(_In_ HINSTANCE hInstance,
...
+		return 1;
 	}
 	else
 	{
-		MessageBox(NULL, L"Unrecognized parameter.\n\nValid options:\n  /START <path>\n  /SINGLE <path>\n  /TASK <name>\n /REGISTER [USER/ALL]\n  /UNREGISTER [USER/ALL]", MB_TITLE, MB_OK);
-		return 1;
+		StartCmder(cmderOptions.cmderStart, cmderOptions.cmderSingle, cmderOptions.cmderTask, cmderOptions.cmderCfgRoot);
 	}
-
+
```
Comment (MartiUK) [inline launcher/src/CmderLauncher.cpp]: Unintentional whitespace here

**Comment by MartiUK** [comment]
@daxgames Ignore me, I didn't specify `%userprofile%\cmder_config` when running with /C, should there be a default location?

**Comment by daxgames** [comment]
@MartiUK I could clean those comments and whitespace up and add a default for. `/c`.  I thought about that after it was done but just wanted it to be done.

**Comment by Stanzilla** [comment]
Yeah a default is needed, bit confusing otherwise, could also make it error if nothing gets specified?

**Comment by MartiUK** [comment]
@Stanzilla @daxgames Well, at the moment cmder does not start at all when running with /C but without an argument.

**Comment by Stanzilla** [comment]
yeah, that should throw an error, imo. I hate it when stuff silently dies and I have no idea why (Hint UWP apps HINT)

**Comment by daxgames** [comment]
@Stanzilla / @MartiUK - I am not a C++ guy so I am struggling with the last mile!

I think if we are going to have a default for the `/c` arg I feel like if `/c` is found in the args we need to pattern match the next argument if it is not NULL to make sure it is either a local or a unc path or just another argument beginning with a `/` so the command line parser knows what to do with the next argument.  Ideas based on whats in the PR?

File: launcher/src/CmderLauncher.cpp (hunk 2 above)
Comment (daxgames) [reply launcher/src/CmderLauncher.cpp]: Fixed!

File: launcher/src/CmderLauncher.cpp (hunk 1 above)
Comment (daxgames) [reply launcher/src/CmderLauncher.cpp]: Fixed!

File: launcher/src/CmderLauncher.cpp (hunk 3 above)
Comment (daxgames) [reply launcher/src/CmderLauncher.cpp]: Fixed!

**Comment by DRSDavidSoft** [comment]
@daxgames I can rewrite the C++ argument parsing logic similar to what I did [here](https://github.com/DRSDavidSoft/SerialUART/blob/5bc2fde286f04a96ce1d96bd5ee3121bc44f0e22/SendCommand.c#L162) in my project.

**Comment by daxgames** [comment]
I jumped the gun still working on it.

**Comment by Stanzilla** [comment]
Really happy with this so far 👍

**Comment by MartiUK** [comment]
Yeah this looks good to me, is there nothing else to add @daxgames?

**Comment by Stanzilla** [comment]
Grats on shipping! Also have fun with closing all the issues that are now fixed :)

**Comment by wireddown** [comment]
For users that need to run `user_profile.ps1` and friends in `profile.d` for **Administrator** sessions in **Powershell**, you cannot use `/c [user cmder root]` because Powershell sessions aren't created from `init.bat`, but instead by

```batch
*PowerShell -ExecutionPolicy Bypass -NoLogo -NoProfile -NoExit -Command "Invoke-Expression 'Import-Module ''%ConEmuDir%\..\profile.ps1'''"
```

and `''%ConEmuDir%\..\profile.ps1''` doesn't accept any command line arguments.

Instead, on the **Environment** panel, set these two variables to match your configuration:

```batch
set CMDER_USER_BIN=C:\full\path\to\your\cmder\bin
set CMDER_USER_CONFIG=C:\full\path\to\your\cmder\config
```

and new Powershell as Admin sessions will use your `user_profile.ps1` script.

**Pull request merged**